Introduction to WebC

darthmall opened the pull request:
A cursory — but not brief — introduction to WebC and Eleventy.

Commit:
First draft: "Introduction to WebC"

darthmall commented:
@5t3ph I tried to keep this short, but it is still pretty long. Let me know if you’d like me to trim it down and we can talk about what is superfluous or just verbose.

I’ve left the PR as a draft because I want to sleep on it and re-read it tomorrow or the day after with fresh eyes and make any changes that jump out at me, but I wanted to share the draft with you so that if you have any changes you’d like to see, I can get those in before or during any final polishing I wanted to put in.

5t3ph commented on src/posts/introduction-webc.md:
````diff
@@ -0,0 +1,236 @@
+
+So if we have a file — `page.webc` — with all of our HTML boilerplate, we insert our footer into that file using this custom element.
+
+```html
````
Let's add comments in the code sample to emphasize the template since we can assume `.webc` is an unfamiliar extension and it otherwise looks like an HTML file. (and we know folks skim and may not see the name in the preceding paragraph 😉 )

```
<!-- page.webc -->
```

5t3ph commented on src/posts/introduction-webc.md:
````diff
@@ -0,0 +1,236 @@
+</style>
+```
+
+One advantage to writing your components this way instead of having your markup and styles in separate files is that when you’re markup changes, it’s easier to remember and update the styles. If we introduce a `<div>` to our footer to wrap the paragraph at some maximum width, our (admittedly contrived) `font-family` declaration will no longer apply to the copyright notice in the footer. Being abl ...
````
```suggestion
One advantage to writing your components this way instead of having your markup and styles in separate files is that when you’re making changes, it’s easier to remember and update the styles. If we introduce a `<div>` to our footer to wrap the paragraph at some maximum width, our (admittedly contrived) `font-family` declaration will no longer apply to the copyright notice in the footer. Being able to see the markup and styles right next to each other helps us keep them synchronized.
```

5t3ph commented on src/posts/introduction-webc.md:
```diff
@@ -0,0 +1,236 @@
+
+One advantage to writing your components this way instead of having your markup and styles in separate files is that when you’re markup changes, it’s easier to remember and update the styles. If we introduce a `<div>` to our footer to wrap the paragraph at some maximum width, our (admittedly contrived) `font-family` declaration will no longer apply to the copyright notice in the footer. Being abl ...
+
+> When you have a component that is just HTML, WebC will replace the custom element with that HTML. But as soon as you introduce either a `<script>` tag or a `<style>` tag to the component, WebC will, by default, leave the custom element in place in case you are using it in either your scripts or your styles. You can tell WebC to remove the custom element by adding [webc:nokeep](https://www.11ty. ...
```
Maybe switch this out of blockquote and add a headline like `### Custom element tag rendering` or something like that?

5t3ph commented on src/posts/introduction-webc.md:
```diff
@@ -0,0 +1,236 @@
+
+When WebC processes your components and it finds `<style>` or `<script>` tags in them, it can extract those assets into bundles for you to use later. This allows you to collect your assets into a single `<style>` tag inside the page `<head>`, rather than leaving `<style>` tags strewn about your HTML. See the Eleventy docs on WebC for information on [how to include bundled assets in your final mar ...
+
+> By default, when using the Eleventy plugin for WebC, asset bundling is turned on. If you are using WebC directly, however, bundling is off by default.
```
```suggestion
> By default, when using the Eleventy plugin for WebC, asset bundling is turned on. If you are [using WebC directly](https://github.com/11ty/webc), however, bundling is off by default.
```

5t3ph commented on src/posts/introduction-webc.md:
````diff
@@ -0,0 +1,236 @@
+  font-family: fantasy;
+}
+```
+
````
```suggestion
> **Did you know**: [`:host` is a native CSS feature](https://developer.mozilla.org/en-US/docs/Web/CSS/:host) that is usually in reference to the shadow DOM root host of web components. WebC is using `:host` a hook to override with the generated scoped class.
```

5t3ph commented on src/posts/introduction-webc.md:
````diff
@@ -0,0 +1,236 @@
+</style>
+```
+
+Here we’ve added the `webc:scoped` attribute to our `<style>` element, and introduced the `:host` pseudo-class to our selectors. WebC now knows to generate a unique class for each `<site-footer>` it processes and to replace `:host` with a class selector matching this generated class name. So our HTML will end up looking something like:
````
```suggestion
Here we’ve added the `webc:scoped` attribute to our `<style>` element, and introduced the `:host` pseudo-class to our selectors. WebC now knows to generate a unique class for `<site-footer>`, which will replace `:host` with a class selector matching this generated class name. The generated class will be shared across `<site-footer>` instances. So our HTML will end up looking something like:
```

5t3ph commented on src/posts/introduction-webc.md:
```diff
@@ -0,0 +1,236 @@
+
+WebC supports the use of `<slot></slot>` for inserting markup inside your components when they’re used. And as with templates and slots for web components, you can define multiple named slots; something that none of the template languages Eleventy currently supports can do.
+
+To illustrate how this can be useful, let’s create a card component that needs to have a title, some media, and a brief description. We won’t concern ourselves with the CSS for this component, but let’s assume we’re doing some fancy things that require us to place the aformentione title, media, and description inside some wrappers. We can define `a-card.webc` like this:
```
Love this example! It really emphasizes how WebC is useful for static components 🙌

5t3ph commented on src/posts/introduction-webc.md:
````diff
@@ -0,0 +1,236 @@
+  }
+</style>
+```
+
````
Include a blockquote callout here to note that the class format and `connectedCallback()` are native web component constructs and link to MDN docs that best fit perhaps?

5t3ph commented on src/posts/introduction-webc.md:
````diff
@@ -0,0 +1,236 @@
+</style>
+```
+
+Now we can put `<theme-select></theme-select>` somewhere in our site, and we’ll end up with
````
```suggestion
Now we can put `<theme-select></theme-select>` somewhere in our site, and we’ll end up with the following in our final, rendered markup:
```

... and delete the hanging part of the sentence below the sample

5t3ph commented on src/posts/introduction-webc.md:
```diff
@@ -0,0 +1,236 @@
+
+> There are plans to allow you to [aggregate arbitrary HTML](https://github.com/11ty/webc/issues/36) from your WebC components the way that styles and scripts are aggregated. This would mean that you could include a `<template>` as part of your component and guarantee that in the final markup there was only one copy of the template regardless of how many instances of the component you placed on t ...
+
+## Learn more
```
```suggestion
## Learn more about WebC
```

5t3ph reviewed (changes requested):
@darthmall this is excellent, no issues on the length as I think it's a fabulous overview of exactly the kind of thing I had questions on and really exposes some practical use cases without being too heavy. Thank you so much for your time on this!!

Commit:
Second draft

Included revisions from Steph’s comments in the PR, as well as a few
small rewrites of my own.

darthmall commented:
@5t3ph I think I got all your notes in — thanks for the feedback. I made a few small rewrites myself, after a re-read. Let me know what you think.

5t3ph commented:
@darthmall looks excellent! If you are ready to switch it out of draft, I'll approve it and get it merged 🎉
